### ISSUE TITLE
SDK 44 upgrade - basic map marker functionality is broken (iOS/Android)

### ISSUE THREAD

**Issue by nakedgun** [expo/expo#16061]
### Summary

This case refers to basic functionality of map markers which has broken since upgrading to SDK 44. These issues do not reproduce on SDK 42/43 (43 sharing the same RN version) nor with a vanilla RN setup. This leads me to conclude this is a expo specific bug - hence why I'm posting it here and not in the `react-native-maps` repo.

**Steps:**

1. Add a map using `react-native-maps`.
2. Programmatically add some markers
3. Programmatically remove some markers, or update their pin color in response to a click.

**result:** markers do not get removed from the map - and pin colors do not update.

**SDK 43 (working):**

`react-native-maps@0.28.1`

![sdk_43](https://user-images.githubusercontent.com/5262234/151040512-2f123b26-9c8b-42e7-ab78-0e59fa884754.gif)

**SDK 44 (broken):**

`react-native-maps@0.29.4`

![sdk_44](https://user-images.githubusercontent.com/5262234/151040533-842663cf-f81c-4231-ba7d-5204ac94b11f.gif)

**Notes:** 

- changing the react-native-maps version has no effect - e.g. SDK 43 works fine with `0.28.1` or `0.29.4` - SDK44 doesn't work with either `0.28.1` or `0.29.4`
- adding new markers is OK - but removing or updating existing ones does not work.
- I have tried forcing a manual redraw of the markers - and playing around with static/non-static keys (more below on this).

### Managed or bare workflow? If you have `ios/` or `android/` directories in your project, the answer is bare!

managed

### What platform(s) does this occur on?

Android, iOS

### SDK Version (managed workflow only)

44

### Environment

  Expo CLI 5.0.3 environment info:
    System:
      OS: macOS 12.0.1
      Shell: 5.8 - /bin/zsh
    Binaries:
      Node: 16.0.0 - ~/.nvm/versions/node/v16.0.0/bin/node
      Yarn: 1.22.17 - ~/.yarn/bin/yarn
      npm: 7.10.0 - ~/.nvm/versions/node/v16.0.0/bin/npm
      Watchman: 4.9.0 - /usr/local/bin/watchman
    Managers:
      CocoaPods: 1.9.3 - /Users/jkl/.rvm/rubies/ruby-2.6.5/bin/pod
    SDKs:
      iOS SDK:
        Platforms: DriverKit 21.0.1, iOS 15.0, macOS 12.0, tvOS 15.0, watchOS 8.0
      Android SDK:
        API Levels: 21, 25, 26, 27, 28, 29
        Build Tools: 26.0.1, 26.0.2, 27.0.3, 28.0.2, 28.0.3, 29.0.2
        System Images: android-21 | ARM EABI v7a, android-22 | Google APIs ARM EABI v7a, android-28 | Google Play Intel x86 Atom, android-29 | Google APIs Intel x86 Atom, android-30 | Google APIs Intel x86 Atom, android-30 | Google Play Intel x86 Atom
    IDEs:
      Android Studio: 2020.3 AI-203.7717.56.2031.7935034
      Xcode: 13.1/13A1030d - /usr/bin/xcodebuild
    npmPackages:
      expo: ~44.0.0 => 44.0.5
      react: 17.0.1 => 17.0.1
      react-dom: 17.0.1 => 17.0.1
      react-native: 0.64.3 => 0.64.3
      react-native-web: 0.17.1 => 0.17.1
    npmGlobalPackages:
      expo-cli: 5.0.3
    Expo Workflow: managed

### Reproducible demo

```js
{
  "name": "maps-test",
  "version": "1.0.0",
  "main": "node_modules/expo/AppEntry.js",
  "scripts": {
    "start": "expo start",
    "android": "expo start --android",
    "ios": "expo start --ios",
    "web": "expo start --web",
    "eject": "expo eject"
  },
  "dependencies": {
    "expo": "~44.0.0",
    "expo-status-bar": "~1.2.0",
    "react": "17.0.1",
    "react-dom": "17.0.1",
    "react-native": "0.64.3",
    "react-native-maps": "0.29.4",
    "react-native-web": "0.17.1"
  },
  "devDependencies": {
    "@babel/core": "^7.12.9"
  },
  "private": true
}
```
```js
import React from 'react';
import { Button, Dimensions, Platform, View } from "react-native";
import MapView, { Marker } from 'react-native-maps';
class App extends React.Component {

  state = {
    countMarkers: 0,
    activeMarkerIndex: null,
    markerColor: 'red'
  }

  changeMarkerColors = () => {
    const color = this.state.markerColor === 'red' ? 'purple' : 'red';
    this.setState({
      markerColor: color
    })
  }

  mapMarkers = () => {
    if (this.state.countMarkers === 0) {
      console.log('no markers')
      return null;
    }
    if (this.state.countMarkers === 1) {
      console.log('1 marker')
      return (
        <Marker
          onPress={() => this.setState({activeMarkerIndex: 1})}
          tracksViewChanges={false}
          identifier={'1'}
          key={`${1}-${this.state.activeMarkerIndex === 1 ? '-active' : '-inactive'}-${this.state.markerColor}`}
          coordinate={{
            latitude: 37.335480,
            longitude: -121.893028,
          }}
          pinColor={this.state.activeMarkerIndex === 1 ? 'green': this.state.markerColor}
        />
      )
    } else {
      console.log('2 markers')
      return (
        <>
          <Marker
            tracksViewChanges={false}
            onPress={() => this.setState({activeMarkerIndex: 2})}
            identifier={'2'}
            key={`${2}-${this.state.activeMarkerIndex === 2 ? '-active' : '-inactive'}-${this.state.markerColor}`}
            coordinate={{
              latitude: 39.335480,
              longitude: -121.893028,
            }}
            pinColor={this.state.activeMarkerIndex === 2 ? 'green': this.state.markerColor}
          />
          <Marker
            tracksViewChanges={false}
            onPress={() => this.setState({activeMarkerIndex: 3})}
            identifier={'3'}
            key={`${3}-${this.state.activeMarkerIndex === 3 ? '-active' : '-inactive'}-${this.state.markerColor}`}
            coordinate={{
              latitude: 32.452961,
              longitude: -96.7969879,
            }}
            pinColor={this.state.activeMarkerIndex === 3 ? 'green': this.state.markerColor}
          />
        </>
      )
    }
  };

  render() {
    const markers = this.mapMarkers();
    return (
      <View style={{ flex: 1, alignItems: 'center' }}>
        <MapView
          style={{ position: 'absolute', top: 0, left: 0, right: 0, bottom: 0 }}
          region={{latitude: 38, longitude: -100, latitudeDelta: 60, longitudeDelta: 60 }}
        >
          {markers}
        </MapView>
        <View style={{ position: 'absolute', bottom: 50 }}>
          <Button title={'0 markers'} onPress={() => this.setState({ countMarkers: 0})}></Button>
          <Button title={'1 markers'} onPress={() => this.setState({ countMarkers: 1})}></Button>
          <Button title={'2 markers'} onPress={() => this.setState({ countMarkers: 2})}></Button>
          <Button title={'Change inactive marker color'} onPress={() => this.changeMarkerColors()}></Button>
        </View>
      </View>
    );
  }
}
export default App;
```

**Note:** the demo above uses a dynamic key to force the markers to update. Traditionally, this is the only way to force pinColor updates on Android, though not necessary on iOS. It's worth noting that key quirks between iOS/Android also vary between SDK 43/44:

- SDK 43: using a dynamic key (e.g. `${key}-${isActive}`) is the only way to force pinColor updates on Android. It also works fine on iOS.
- SDK 44: using a dynamic key now breaks color updates on iOS - you need to use a static key for this to work. On Android, you now have to use a **random** key each time for the marker to receive updates (e.g. in color). So before, whilst `${key}-${isActive}` would be enough, now it has to be a key which has never been used before:, e.g. `${key}-${isActive}-${randomInt}`.

**Comment by brentvatne** [comment]
can you share a repo that we can clone and explain in what context you are observing this? is this in expo go? can you explain how you verified that were unable to repro with a "vanilla RN setup"?

**Comment by nakedgun** [comment]
> can you share a repo that we can clone and explain in what context you are observing this? is this in expo go? can you explain how you verified that were unable to repro with a "vanilla RN setup"?

@brentvatne Sure, no problem. For expo, it reproduces in Expo Go (Android/iOS) on simulators and physical devices. I haven't tested a compiled build yet - I can if you want.

[Here's a repo](https://github.com/nakedgun/react-native-maps-bug) which contains demos for SDK43 (working), SDK44 (not working) and vanilla RN (working). Each repo uses the same map component. The vanilla RN example uses the same react-native/react-native-maps versions as SDK 44 and it works fine. That's why I came to the conclusion this is an expo specific issue.

**Comment by brentvatne** [comment]
@nakedgun - i'm not sure what i'm doing wrong but i am unable to reproduce this issue in the repo that you provided

here is the sdk 44 project running in expo go for m

https://user-images.githubusercontent.com/90494/151074503-8396beb1-6dbd-4ed6-9053-6b02fe64b33d.mov

.

**Comment by nakedgun** [comment]
@brentvatne I uninstalled/reinstalled expo go and it now works for me. As far as I can tell, there are two issues here:

1. In my production app, the actual bug with the map is caused by using an older version of `@react-navigation/stack` (`5.14.9`) with expo SDK44 **and** running the map view in a modal stack. Updating `@react-navigation/stack` fixes that.
2. The more confusing issue came from switching between the two apps. If I launched my production app (which has the bug caused by react-navigation) in expo go - then switch to the test repo, the bug still happens. The only way I could get the test repo to work was by uninstalling expo go. No clue why - I'm assuming some kind of weird caching/state pollution going on.

Anyway, a hard lesson learnt. Gonna close this - but thanks for the help!
